File: patterns/ruby/Patterns.rb
def pyramid(n)
  (1..n).each do |i|
    puts ' ' * (n - i) + '*' * (2 * i - 1)
  end
end

def inverted(n)
  n.downto(1) do |i|
    puts ' ' * (n - i) + '*' * (2 * i - 1)
  end
end

def diamond(n)
  pyramid(n)
  (n - 1).downto(1) do |i|
    puts ' ' * (n - i) + '*' * (2 * i - 1)
  end
end

def hollow_pyramid(n)
  (1..n).each do |i|
    if i == 1
      puts ' ' * (n - 1) + '*'
    elsif i == n
      puts '*' * (2 * n - 1)
    else
      puts ' ' * (n - i) + '*' + ' ' * (2 * i - 3) + '*'
    end
  end
end

def pascals(n)
  row = [1]
  n.times do
    row_str = row.join(' ')
    puts row_str.center(2 * n)
    next_row = [1]
    (row.length - 1).times do |i|
      next_row.push(row[i] + row[i + 1])
    end
    next_row.push(1)
    row = next_row
  end
end

def hourglass(n)
  inverted(n)
  (2..n).each do |i|
    puts ' ' * (n - i) + '*' * (2 * i - 1)
  end
end

def zigzag(n)
    height = 4
    n.times do |i|
        height.times do |j|
            if j == 0 || j == height - 1
                puts '*' * n
            else
                puts ' ' * (n - i - 1) + '*'
            end
        end
    end
end

def main
  pattern = ARGV[0] || 'pyramid'
  n = ARGV[1]&.to_i || 5

  funcs = {
    'pyramid' => method(:pyramid),
    'inverted' => method(:inverted),
    'diamond' => method(:diamond),
    'hollow_pyramid' => method(:hollow_pyramid),
    'pascals' => method(:pascals),
    'hourglass' => method(:hourglass),
    'zigzag' => method(:zigzag)
  }

  if funcs.key?(pattern)
    funcs[pattern].call(n)
  else
    puts "Unknown pattern. Choose from: #{funcs.keys.join(', ')}"
  end
end

main

File: patterns/cpp/Patterns.cpp
#include <iostream>
#include <string>
#include <vector>
#include <functional>
#include <map>

void printChars(char c, int n) {
    if (n > 0) {
        std::cout << std::string(n, c);
    }
}

void pyramid(int n) {
    for (int i = 1; i <= n; ++i) {
        printChars(' ', n - i);
        printChars('*', 2 * i - 1);
        std::cout << std::endl;
    }
}

void inverted(int n) {
    for (int i = n; i >= 1; --i) {
        printChars(' ', n - i);
        printChars('*', 2 * i - 1);
        std::cout << std::endl;
    }
}

void diamond(int n) {
    pyramid(n);
    for (int i = n - 1; i >= 1; --i) {
        printChars(' ', n - i);
        printChars('*', 2 * i - 1);
        std::cout << std::endl;
    }
}

void hollowPyramid(int n) {
    for (int i = 1; i <= n; ++i) {
        printChars(' ', n - i);
        if (i == 1 || i == n) {
            printChars('*', 2 * i - 1);
        } else {
            std::cout << "*";
            printChars(' ', 2 * i - 3);
            std::cout << "*";
        }
        std::cout << std::endl;
    }
}

void pascals(int n) {
    std::vector<long> row = {1};
    for (int i = 0; i < n; ++i) {
        std::string rowStr;
        for (size_t k = 0; k < row.size(); ++k) {
            rowStr += std::to_string(row[k]) + (k == row.size() - 1 ? "" : " ");
        }

        int padding = (2 * n - rowStr.length()) / 2;
        printChars(' ', padding);
        std::cout << rowStr << std::endl;

        std::vector<long> nextRow;
        nextRow.push_back(1);
        for (size_t j = 0; j < row.size() - 1; ++j) {
            nextRow.push_back(row[j] + row[j+1]);
        }
        nextRow.push_back(1);
        row = nextRow;
    }
}

void hourglass(int n) {
    inverted(n);
    for (int i = 2; i <= n; ++i) {
        printChars(' ', n - i);
        printChars('*', 2 * i - 1);
        std::cout << std::endl;
    }
}

void zigzag(int n) {
    int height = 4;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < height; ++j) {
            if (j == 0 || j == height - 1) {
                printChars('*', n);
                std::cout << std::endl;
            } else {
                printChars(' ', n - i - 1);
                std::cout << "*" << std::endl;
            }
        }
    }
}

int main(int argc, char* argv[]) {
    std::string pattern = "pyramid";
    int n = 5;

    if (argc >= 2) {
        pattern = argv[1];
    }
    if (argc >= 3) {
        try {
            n = std::stoi(argv[2]);
        } catch (const std::invalid_argument& e) {
            std::cerr << "Error: Size must be an integer." << std::endl;
            return 1;
        }
    }

    std::map<std::string, std::function<void(int)>> funcs = {
        {"pyramid", pyramid},
        {"inverted", inverted},
        {"diamond", diamond},
        {"hollow_pyramid", hollowPyramid},
        {"pascals", pascals},
        {"hourglass", hourglass},
        {"zigzag", zigzag}
    };

    if (funcs.count(pattern)) {
        funcs[pattern](n);
    } else {
        std::cout << "Unknown pattern. Choose from: pyramid, inverted, diamond, hollow_pyramid, pascals, hourglass, zigzag" << std::endl;
    }

    return 0;
}